pkg/publisher/publisher.go

```diff
@@ -0,0 +1,446 @@
+			_ = os.RemoveAll(tmpDir)
+			return "", fmt.Errorf("writing car chunk to file: %v", err)
+		}
+		c++
+	}
+	return tmpDir, nil
+}
+
+// CarChunksToNftStorage publishes the car chunks in the specified dir to nft.storage.
+func (p *Publisher) CarChunksToNftStorage(
+	ctx context.Context,
+	tmpDir string,
+	concurrency int,
+	rateLimit time.Duration,
+) error {
+	chunks, err := os.ReadDir(tmpDir)
+	if err != nil {
+		return fmt.Errorf("reading tmp dir: %v", err)
+	}
+
+	var jobs1 []wpool.Job
+	for i, chunk := range chunks {
+		jobs1 = append(jobs1, wpool.Job{
+			ID:   wpool.JobID(i),
+			Desc: fmt.Sprintf("upload chunk %d", i),
+			ExecFn: func(de fs.DirEntry) wpool.ExecutionFn {
+				return func(ctx context.Context) (interface{}, error) {
+					f, err := os.Open(path.Join(tmpDir, de.Name()))
+					if err != nil {
+						return nil, fmt.Errorf("opening chunk file: %v", err)
+					}
+					res, err := p.carStorage.PutCar(ctx, f)
+					if err != nil {
+						return nil, fmt.Errorf("uploading car chunk: %v", err)
+					}
+					return res, nil
+				}
+			}(chunk),
+		})
+	}
+
+	ctx1, cancel1 := context.WithCancel(ctx)
+	defer cancel1()
+
+	p1 := wpool.New(concurrency, rate.Every(rateLimit))
+	go p1.GenerateFrom(jobs1)
+	go p1.Run(ctx1)
+	for r := range p1.Results() {
+		if r.Err != nil {
+			return fmt.Errorf("executing job %d, %s: %v", r.ID, r.Desc, r.Err)
+		}
+		log.Default().Printf("processed job %d: %s, %v", r.ID, r.Desc, r.Value)
+	}
+	_ = os.RemoveAll(tmpDir)
+	return nil
+}
+
+// CidToCarStorage writes a car file from a cid and uploads it to car storage service.
+func (p *Publisher) CidToCarStorage(ctx context.Context, c cid.Cid) (cid.Cid, error) {
+	carReader, carWriter := io.Pipe()
+
+	jobs := []wpool.Job{
+		{
+			ID:   1,
+			Desc: "write car",
+			ExecFn: func(ctx context.Context) (interface{}, error) {
+				if err := car.WriteCar(ctx, p.ipfsClient.Dag(), []cid.Cid{c}, carWriter); err != nil {
+					return nil, fmt.Errorf("writing car: %v", err)
+				}
+				_ = carWriter.Close()
+				return nil, nil
+			},
+		},
+		{
+			ID:   2,
+			Desc: "upload",
+			ExecFn: func(ctx context.Context) (interface{}, error) {
+				c, err := p.carStorage.PutCar(ctx, carReader)
+				if err != nil {
+					return nil, fmt.Errorf("uploading car: %v", err)
+				}
+				return c, nil
+			},
+		},
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	pool := wpool.New(2, rate.Inf)
+	go pool.GenerateFrom(jobs)
+	go pool.Run(ctx)
+	var res cid.Cid
+	for r := range pool.Results() {
+		if r.Err != nil {
+			return cid.Cid{}, fmt.Errorf("executing job %d, %s: %v", r.ID, r.Desc, r.Err)
+		}
+		if r.ID == 2 {
+			res = r.Value.(cid.Cid)
+		}
+	}
+	log.Default().Printf("%s uploaded to car storage service", res.String())
+	return res, nil
+}
+
+// RigsIndexToCarStorage creates a dagcbor index from the provided Rigs and adds it to car storage service.
+func (p *Publisher) RigsIndexToCarStorage(ctx context.Context, rigs []local.Rig) (cid.Cid, error) {
```

Comment:
Goal of this func is to create a car file that maps each rig id (`int`) to a cid link for imagery that was already added to web3.storage/nft.storage in a previous step.